File: tests/benchmarks/comprehensive_benchmark.py
#!/usr/bin/env python3

# Comprehensive Orus Language Benchmark - Python Version
# Tests all currently supported features in one complex benchmark

import time

print("=== Orus Comprehensive Performance Benchmark ===")

start_time = int(time.time() * 1000000)  # microseconds

# === ARITHMETIC OPERATIONS INTENSIVE TEST ===
print("Phase 1: Intensive Arithmetic Operations")

# Complex arithmetic chains
a = 100
b = 50
c = 25
d = 12
e = 6

# Multi-step calculations
result1 = a + b * c - d // e
result2 = (a - b) * (c + d) // e
result3 = a * b // c + d - e
result4 = (a + b + c + d + e) // (a - b - c - d - e)

# Nested arithmetic expressions
complex1 = ((a + b) * c - d) // ((e + d) * c - b)
complex2 = (a * b + c * d) // (a - b + c - d)
complex3 = ((a // b) * c + d) - ((e * d) // c + b)

# Iterative calculations (simulating loops)
sum_val = 0
counter = 0
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter

print("Arithmetic Phase Results:")
print("Complex result 1:", complex1)
print("Complex result 2:", complex2)
print("Complex result 3:", complex3)
print("Iterative sum:", sum_val)

# === VARIABLE OPERATIONS INTENSIVE TEST ===
print("Phase 2: Intensive Variable Operations")

# Variable assignment chains
var1 = 10
var2 = var1 * 2
var3 = var2 + var1
var4 = var3 - var2
var5 = var4 * var1
var6 = var5 // var2
var7 = var6 + var3
var8 = var7 - var4
var9 = var8 * var5
var10 = var9 // var6

# Variable swapping simulation
temp_a = var1
temp_b = var2
var1 = temp_b
var2 = temp_a

# Complex variable interdependencies
base = 5
multiplier = 3
offset = 7
threshold = 20

calc_a = base * multiplier + offset
calc_b = calc_a - threshold
calc_c = calc_b * base // multiplier
calc_d = calc_c + offset - threshold
calc_e = calc_d * multiplier // base

# Variable reuse patterns
accumulator = 0
accumulator = accumulator + base
accumulator = accumulator * multiplier
accumulator = accumulator - offset
accumulator = accumulator + threshold
accumulator = accumulator // base
accumulator = accumulator * offset
accumulator = accumulator - multiplier

print("Variable Phase Results:")
print("Final var10:", var10)
print("Swapped var1:", var1)
print("Swapped var2:", var2)
print("Calc chain result:", calc_e)
print("Accumulator result:", accumulator)

# === LITERAL VALUES INTENSIVE TEST ===
print("Phase 3: Intensive Literal Operations")

# Large number calculations
big1 = 1000000
big2 = 500000
big3 = 250000
big4 = 125000
big5 = 62500

# Operations with large numbers
large_sum = big1 + big2 + big3 + big4 + big5
large_diff = big1 - big2 - big3 - big4 - big5
large_prod = big1 // 1000 * big2 // 1000
large_div = big1 // big2 * big3 // big4

# Mixed literal and variable operations
mixed1 = 42 + var1 * 17 - 8
mixed2 = 99 - var2 // 3 + 21
mixed3 = 77 * var3 + 13 - 56
mixed4 = 88 // var4 - 44 + 33

# Decimal-like operations (using integer division)
decimal1 = 100 * 355 // 113
decimal2 = 1000 * 22 // 7
decimal3 = 10000 * 618 // 1000

print("Literal Phase Results:")
print("Large sum:", large_sum)
print("Large difference:", large_diff)
print("Large product:", large_prod)
print("Mixed calculation 1:", mixed1)
print("Mixed calculation 2:", mixed2)
print("Decimal approximation 1:", decimal1)
print("Decimal approximation 2:", decimal2)

# === TIMESTAMP OPERATIONS INTENSIVE TEST ===
print("Phase 4: Intensive Timestamp Operations")

# Multiple timestamp measurements
ts1 = int(time.time() * 1000000)
ts2 = int(time.time() * 1000000)
ts3 = int(time.time() * 1000000)
ts4 = int(time.time() * 1000000)
ts5 = int(time.time() * 1000000)

# Timestamp arithmetic
ts_diff1 = ts2 - ts1
ts_diff2 = ts3 - ts2
ts_diff3 = ts4 - ts3
ts_diff4 = ts5 - ts4

# Complex timestamp calculations
ts_total = ts_diff1 + ts_diff2 + ts_diff3 + ts_diff4
ts_avg = ts_total // 4
ts_max = ts_diff1
temp_check = ts_diff2 - ts_max
temp_check = ts_diff3 - ts_max
temp_check = ts_diff4 - ts_max

# Timestamp with arithmetic operations
ts_calc1 = ts1 + 1000
ts_calc2 = ts2 * 2 // 2
ts_calc3 = ts3 - 500 + 500
ts_calc4 = (ts4 + ts5) // 2

print("Timestamp Phase Results:")
print("Timestamp 1:", ts1)
print("Timestamp 5:", ts5)
print("Total time diff:", ts_total)
print("Average time diff:", ts_avg)
print("Complex timestamp calc:", ts_calc4)

# === PRINT OPERATIONS INTENSIVE TEST ===
print("Phase 5: Intensive Print Operations")

# Multiple print statements with different data
print("Testing multiple print operations:")
print("Number:", 42)
print("Calculation:", 10 + 5)
print("Variable:", accumulator)
print("Expression:", (a + b) * c)
print("Large number:", big1)
print("Time:", ts1)

# Print with complex expressions
print("Complex expression 1:", ((100 + 200) * 3) // 2)
print("Complex expression 2:", (500 - 100) * 2 + 50)
print("Complex expression 3:", 1000 // 10 + 200 // 4)

# === COMPREHENSIVE INTEGRATION TEST ===
print("Phase 6: Comprehensive Integration")

# Combine all features in complex calculations
integration_start = int(time.time() * 1000000)

# Complex integrated calculation
step1 = (big1 // 1000) + (var1 * var2)
step2 = step1 - (complex1 * 10)
step3 = step2 + (ts_avg * 100)
step4 = step3 // (accumulator + 1)
step5 = step4 * (decimal1 // 100)

# Multiple variable interactions
chain_a = step1 + step2
chain_b = step3 + step4
chain_c = step5 + chain_a
chain_d = chain_b + chain_c
final_result = chain_d // 4

# Final timestamp
integration_end = int(time.time() * 1000000)
integration_time = integration_end - integration_start

print("Integration Phase Results:")
print("Step 1:", step1)
print("Step 2:", step2)
print("Step 3:", step3)
print("Step 4:", step4)
print("Step 5:", step5)
print("Final integrated result:", final_result)
print("Integration time:", integration_time)

# === FINAL BENCHMARK RESULTS ===
end_time = int(time.time() * 1000000)
total_elapsed = end_time - start_time

print("=== BENCHMARK COMPLETE ===")
print("Total execution time:", total_elapsed)
print("Operations completed: 200+")
print("Variables used: 50+")
print("Arithmetic operations: 100+")
print("Print statements: 40+")
print("Timestamp operations: 20+")
print("Final benchmark score:", total_elapsed + final_result)
print("=== Orus Comprehensive Benchmark Complete ===")

File: tests/benchmarks/ComprehensiveBenchmark.java
/**
 * Java Comprehensive Performance Benchmark
 * Equivalent to comprehensive_benchmark.orus for fair comparison
 */
public class ComprehensiveBenchmark {
    
    public static void main(String[] args) {
        System.out.println("=== Java Comprehensive Performance Benchmark ===");
        
        long startTime = System.nanoTime();
        
        // === ARITHMETIC OPERATIONS INTENSIVE TEST ===
        System.out.println("Phase 1: Intensive Arithmetic Operations");
        
        // Complex arithmetic chains
        int a = 100;
        int b = 50;
        int c = 25;
        int d = 12;
        int e = 6;
        
        // Multi-step calculations
        int result1 = a + b * c - d / e;
        int result2 = (a - b) * (c + d) / e;
        int result3 = a * b / c + d - e;
        int result4 = (a + b + c + d + e) / (a - b - c - d - e);
        
        // Nested arithmetic expressions
        int complex1 = ((a + b) * c - d) / ((e + d) * c - b);
        int complex2 = (a * b + c * d) / (a - b + c - d);
        int complex3 = ((a / b) * c + d) - ((e * d) / c + b);
        
        // Iterative calculations (simulating loops)
        int sum = 0;
        int counter = 0;
        for (int i = 0; i < 10; i++) {
            counter = counter + 1;
            sum = sum + counter;
        }
        
        System.out.println("Arithmetic Results:");
        System.out.println("Result 1:" + result1);
        System.out.println("Result 2:" + result2);
        System.out.println("Result 3:" + result3);
        System.out.println("Result 4:" + result4);
        System.out.println("Complex 1:" + complex1);
        System.out.println("Complex 2:" + complex2);
        System.out.println("Complex 3:" + complex3);
        System.out.println("Sum:" + sum);
        
        // === VARIABLE ASSIGNMENT AND COMPUTATION TEST ===
        System.out.println("Phase 2: Variable Operations and Computations");
        
        // Multiple variable assignments and operations
        int var1 = 100;
        int var2 = var1 / 2;
        int var3 = var2 * 3;
        int var4 = var3 - var1;
        int var5 = var4 + var2;
        
        // Chained assignments
        int x = 10;
        int y = x * 2;
        int z = y + x;
        x = z - y;
        y = x + z;
        z = y * x;
        
        // Complex variable interactions
        int alpha = var1 + var2;
        int beta = var3 * var4;
        int gamma = var5 / alpha;
        int delta = beta - gamma;
        int epsilon = alpha + beta + gamma + delta;
        
        System.out.println("Variable Results:");
        System.out.println("Var1:" + var1 + " Var2:" + var2 + " Var3:" + var3);
        System.out.println("X:" + x + " Y:" + y + " Z:" + z);
        System.out.println("Epsilon:" + epsilon);
        
        // === TYPE OPERATIONS AND CASTING TEST ===
        System.out.println("Phase 3: Type Operations");
        
        // Integer operations
        int int_val = 42;
        int int_result = int_val * 2 + 10;
        
        // Mixed arithmetic with type consistency
        int mixed1 = int_val + 100;
        int mixed2 = mixed1 * 2;
        int mixed3 = mixed2 / 3;
        int final_mixed = mixed1 + mixed2 + mixed3;
        
        System.out.println("Type Results:");
        System.out.println("Int result:" + int_result);
        System.out.println("Final mixed:" + final_mixed);
        
        // === EXPRESSION COMPLEXITY TEST ===
        System.out.println("Phase 4: Complex Expression Evaluation");
        
        // Deeply nested expressions
        int nested1 = ((((a + b) * c) - d) / e) + ((((x + y) * z) - var1) / var2);
        int nested2 = (((var3 + var4) * (var5 - alpha)) / ((beta + gamma) - (delta + epsilon)));
        int nested3 = ((alpha * beta) + (gamma * delta)) - ((epsilon * var1) - (var2 * var3));
        
        // Conditional-like calculations using arithmetic
        int cond1 = (a > b) ? 1 : 0;
        int cond2 = (c < d) ? 1 : 0;
        int cond3 = (e == 6) ? 1 : 0;
        int cond_result = cond1 + cond2 + cond3;
        
        System.out.println("Complex Expression Results:");
        System.out.println("Nested 1:" + nested1);
        System.out.println("Nested 2:" + nested2);
        System.out.println("Nested 3:" + nested3);
        System.out.println("Conditional result:" + cond_result);
        
        // === COMPUTATIONAL STRESS TEST ===
        System.out.println("Phase 5: Computational Stress");
        
        // Large computation chains
        int stress_result = 0;
        for (int i = 1; i <= 100; i++) {
            stress_result = stress_result + (i * i) + (i / 2) + (i % 7);
        }
        
        // Mathematical sequences
        int fibonacci = fibonacci(15);
        int factorial = factorial(8);
        int geometric = 1;
        for (int i = 0; i < 10; i++) {
            geometric = geometric * 2;
        }
        
        // Final comprehensive calculation
        int comprehensive_result = result1 + result2 + result3 + result4 + 
                                 complex1 + complex2 + complex3 + sum + 
                                 epsilon + final_mixed + nested1 + nested2 + nested3 + 
                                 stress_result + fibonacci + factorial + geometric;
        
        System.out.println("Stress Results:");
        System.out.println("Stress calculation:" + stress_result);
        System.out.println("Fibonacci result:" + fibonacci);
        System.out.println("Factorial result:" + factorial);
        System.out.println("Geometric result:" + geometric);
        System.out.println("Final comprehensive result:" + comprehensive_result);
        
        long endTime = System.nanoTime();
        long totalTimeMs = (endTime - startTime) / 1_000_000;
        
        System.out.println("Computation time:" + totalTimeMs);
        System.out.println("=== COMPREHENSIVE BENCHMARK COMPLETE ===");
        System.out.println("Total execution time:" + totalTimeMs);
        System.out.println("Phases completed: 5");
        System.out.println("Operations performed: 200+");
        System.out.println("Variables used: 50+");
        System.out.println("Complex expressions: 20+");
        System.out.println("Final benchmark score:" + totalTimeMs);
        System.out.println("=== Java Comprehensive Benchmark Complete ===");
    }
    
    // Helper functions
    static int fibonacci(int n) {
        if (n <= 1) return n;
        return fibonacci(n - 1) + fibonacci(n - 2);
    }
    
    static int factorial(int n) {
        if (n <= 1) return 1;
        return n * factorial(n - 1);
    }
}
